### ISSUE TITLE
FFmpeg / libmp3lame crash while converting from .wav to .mp3 with vibrato

### ISSUE THREAD

**Issue by chitrang200889** [tanersener/mobile-ffmpeg#488]
**Description**
I have integrated mobile-ffmpeg-full-gpl:4.3.1.LTS in my android app. And trying to convert .wav file to .mp3 format with vibrato option and libmp3lame encoder.

```
ffmpegCommand = "-i input.wav " +
                "-af vibrato=f=4 " +
                "-c:a libmp3lame " +
                "-b:a 96k " +
                "-ac 1 " +
                "-ar 44100 " +
                "-y output.mp3"
```

**Expected behavior**
A command should runt without throwing an exception.

**Current behavior**
It is crashing an app, as it is NDK level crash even try-catch is not able to handle it.

**Logs**
```
a.b.com I/mobile-ffmpeg: Loading mobile-ffmpeg.
a.b.com I/mobile-ffmpeg: Loaded mobile-ffmpeg-full-gpl-arm64-v8a-4.3.1-lts-20200125.
a.b.com D/mobile-ffmpeg: Callback thread started.
a.b.com I/mobile-ffmpeg: ffmpeg version git-2020-01-25-fd11dd500
a.b.com I/mobile-ffmpeg:  Copyright (c) 2000-2020 the FFmpeg developers
a.b.com I/mobile-ffmpeg:   built with Android (5220042 based on r346389c) clang version 8.0.7 (https://android.googlesource.com/toolchain/clang b55f2d4ebfd35bf643d27dbca1bb228957008617) (https://android.googlesource.com/toolchain/llvm 3c393fe7a7e13b0fba4ac75a01aa683d7a5b11cd) (based on LLVM 8.0.7svn)
a.b.com I/mobile-ffmpeg:   configuration: --cross-prefix=aarch64-linux-android- --sysroot=/files/android-sdk/ndk-bundle/toolchains/llvm/prebuilt/linux-x86_64/sysroot --prefix=/home/taner/Projects/mobile-ffmpeg/prebuilt/android-arm64/ffmpeg --pkg-config=/usr/bin/pkg-config --enable-version3 --arch=aarch64 --cpu=armv8-a --cc=aarch64-linux-android21-clang --cxx=aarch64-linux-android21-clang++ --target-os=android --enable-neon --enable-asm --enable-inline-asm --enable-cross-compile --enable-pic --enable-jni --enable-optimizations --enable-swscale --enable-shared --disable-v4l2-m2m --disable-outdev=v4l2 --disable-outdev=fbdev --disable-indev=v4l2 --disable-indev=fbdev --enable-small --disable-openssl --disable-xmm-clobber-test --disable-debug --enable-lto --disable-neon-clobber-test --disable-programs --disable-postproc --disable-doc --disable-htmlpages --disable-manpages --disable-podpages --disable-txtpages --disable-static --disable-sndio --disable-schannel --disable-securetransport --disable-xlib --disable-cuda --disable-cuvid --disa
a.b.com I/mobile-ffmpeg:   libavutil      56. 38.100 / 56. 38.100
a.b.com I/mobile-ffmpeg:   libavcodec     58. 65.102 / 58. 65.102
a.b.com I/mobile-ffmpeg:   libavformat    58. 35.101 / 58. 35.101
a.b.com I/mobile-ffmpeg:   libavdevice    58.  9.103 / 58.  9.103
a.b.com I/mobile-ffmpeg:   libavfilter     7. 70.101 /  7. 70.101
a.b.com I/mobile-ffmpeg:   libswscale      5.  6.100 /  5.  6.100
a.b.com I/mobile-ffmpeg:   libswresample   3.  6.100 /  3.  6.100
a.b.com W/mobile-ffmpeg: [wav @ 0x7294a86600] Estimating duration from bitrate, this may be inaccurate
a.b.com W/mobile-ffmpeg: Guessed Channel Layout for Input Stream #0.0 : mono
a.b.com I/mobile-ffmpeg: Input #0, wav, from '/data/user/0/a.b.com/cache/creation/input.wav':
a.b.com I/mobile-ffmpeg:   Duration: 
a.b.com I/mobile-ffmpeg: 00:00:07.15
a.b.com I/mobile-ffmpeg: , bitrate: 
a.b.com I/mobile-ffmpeg: 705 kb/s
a.b.com I/mobile-ffmpeg:     Stream #0:0
a.b.com I/mobile-ffmpeg: : Audio: pcm_s16le ([1][0][0][0] / 0x0001), 44100 Hz, mono, s16, 705 kb/s
a.b.com I/mobile-ffmpeg: Stream mapping:
a.b.com I/mobile-ffmpeg:   Stream #0:0 -> #0:0
a.b.com I/mobile-ffmpeg:  (pcm_s16le (native) -> mp3 (libmp3lame))
a.b.com I/mobile-ffmpeg: Press [q] to stop, [?] for help
a.b.com I/mobile-ffmpeg: Output #0, mp3, to '/data/user/0/a.b.com/cache/creation/output.mp3':
a.b.com I/mobile-ffmpeg:   Metadata:
a.b.com I/mobile-ffmpeg:     TSSE            : 
a.b.com I/mobile-ffmpeg: Lavf58.35.101
a.b.com I/mobile-ffmpeg:     Stream #0:0
a.b.com I/mobile-ffmpeg: : Audio: mp3 (libmp3lame), 44100 Hz, mono, fltp, 96 kb/s
a.b.com I/mobile-ffmpeg:     Metadata:
a.b.com I/mobile-ffmpeg:       encoder         : 
a.b.com I/mobile-ffmpeg: Lavc58.65.102 libmp3lame
a.b.com I/mobile-ffmpeg: --------- beginning of crash
a.b.com A/libc: psymodel.c:576: void calc_energy(const PsyConst_CB2SB_t *, const FLOAT *, FLOAT *, FLOAT *, FLOAT *): assertion "el >= 0" failed
a.b.com A/libc: Fatal signal 6 (SIGABRT), code -6 (SI_TKILL) in tid 25800 (a.b.com), pid 25800 (a.b.com)
```

**Crash**

```
--------- beginning of crash
 A/libc: psymodel.c:576: void calc_energy(const PsyConst_CB2SB_t *, const FLOAT *, FLOAT *, FLOAT *, FLOAT *): assertion "el >= 0" failed

? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
? A/DEBUG: Build fingerprint: 'samsung/star2qltecs/star2qltecs:10/QP1A.190711.020/G965WVLS7DTE1:user/release-keys'
? A/DEBUG: Revision: '14'
? A/DEBUG: ABI: 'arm64'
? A/DEBUG: Timestamp: 2020-06-29 15:13:17-0400
? A/DEBUG: pid: 1849, tid: 1849, name: a.b.com  >>> a.b.com <<<
? A/DEBUG: uid: 12171
? A/DEBUG: signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
? A/DEBUG: Abort message: 'psymodel.c:576: void calc_energy(const PsyConst_CB2SB_t *, const FLOAT *, FLOAT *, FLOAT *, FLOAT *): assertion "el >= 0" failed'
? A/DEBUG:     x0  0000000000000000  x1  0000000000000739  x2  0000000000000006  x3  0000007fd65f7bb0
? A/DEBUG:     x4  0000000000000000  x5  0000000000000000  x6  0000000000000000  x7  0000000000000008
? A/DEBUG:     x8  00000000000000f0  x9  7f96d7a39856d151  x10 0000000000000001  x11 0000000000000000
? A/DEBUG:     x12 fffffff0fffffbdf  x13 000000005efa3d4c  x14 001c23c1a79207f5  x15 000079d970d48db2
? A/DEBUG:     x16 00000073e009e8c0  x17 00000073e007afe0  x18 00000073e492c000  x19 0000000000000739
? A/DEBUG:     x20 0000000000000739  x21 00000000ffffffff  x22 0000007fd65fc44c  x23 0000007fd65f8640
? A/DEBUG:     x24 0000007fd65fd120  x25 0000007fd65fd3a8  x26 0000007fd65f8240  x27 0000007fd65f9e40
? A/DEBUG:     x28 00000071f9f60900  x29 0000007fd65f7c50
? A/DEBUG:     sp  0000007fd65f7b90  lr  00000073e002c27c  pc  00000073e002c2a8
? A/DEBUG: backtrace:
? A/DEBUG:       #00 pc 00000000000832a8  /apex/com.android.runtime/lib64/bionic/libc.so (abort+160) (BuildId: 55ce0a7d78144b0290f9746ed1615719)
? A/DEBUG:       #01 pc 00000000000839e8  /apex/com.android.runtime/lib64/bionic/libc.so (__assert2+36) (BuildId: 55ce0a7d78144b0290f9746ed1615719)
? A/DEBUG:       #02 pc 0000000000969c60  /data/app/a.b.com-jXqE8oxytEkfSsn6pcdloQ==/lib/arm64/libavcodec.so
```

**Environment**
 - Platform: Android
 - Architecture: arm-v7a, arm-v7a-neon, arm64-v8a
 - Version: mobile-ffmpeg-full-gpl:4.3.1.LTS

**Other**
I referred #238 and a [ticket](https://trac.ffmpeg.org/ticket/8399) to understand the problem but could not find a solution.

**Comment by tanersener** [comment]
What is the connection between this issue and the [ticket](https://trac.ffmpeg.org/ticket/8399)?

**Comment by tanersener** [comment]
This bug is coming from `ffmpeg` and `lame`. You can reproduce it by building `ffmpeg` from `master` using `lame v3.100`. I managed to reproduce it on Mac with `ffmpeg` version `N-98424-g73757577cb`.

Unfortunately, I don't think it will be possible to provide a `mobile-ffmpeg` specific fix for it. It has to be fixed by `ffmpeg` or `lame`. I suggest creating a ticket for it.

**Comment by chitrang200889** [comment]
Thanks for trying it out and giving confirmation. I will report it on the FFmpeg forum.

However, I was just wondering if is there a way to handle this exception?

**Comment by tanersener** [comment]
As far as I know, it is not possible to handle this type of native errors.

**Comment by alexcohn** [comment]
I believe this teaches us that `-DNDEBUG` is not added to CFLAGS for **lame**. Digging just a little deeper, this `-DNDEBUG` probably should be added to `OPTIMIZATION_FLAGS` in **get_cflags()** for **android-common.sh**, **ios-common.sh**, and **tvos-common.sh**.

This should improve performance in some cases, but I wonder how will the resulting library behave for the same **.wav** input with **lame**. Will it still crash, or silently produce an unpleasant **mp3**?

**Comment by tanersener** [comment]
I'm not sure that setting `NDEBUG` is a good idea. `NDEBUG` is disabled in `configure.in` with the following notice. 
```
dnl	Comment out the NDEBUG on release build, they check for issues
dnl	which should also be rejected in a release build.
dnl	The real solution would be to do user friendly error messages
dnl	instead of a core dump. This is a quick and user-unfriendly fix.
dnl		AC_DEFINE(NDEBUG, 1, no debug build)
```

Also, https://sourceforge.net/p/lame/bugs/491 includes more detail about `NDEBUG` and why it shouldn't be used in production.

**Comment by alexcohn** [comment]
That's a cool discussion you found! Now, the challenge is even harder. The existing behavior is probably acceptable in ffmpeg because when **lame** crashes due to garbage input, this is only the current ffmpeg session. With **mobile-ffmpeg**, the whole use app crashes, which is a whole different user experience.

As Alexander Leidinger [recently explained](https://sourceforge.net/p/lame/bugs/491/#82db/a8be/eebf), one cannot simply skip the `assert`s in **lame**. What can probably be done, is redefine their behavior and cancel current execution instead of crashing the app. But if this is not reported too often (whatever your definition of _too often_), this fix can wait.

You still may gain from enabling `NDEBUG` for other libraries. Note that this is a standard behavior for all build systems, including NDK and XCode.

**Comment by tanersener** [comment]
> That's a cool discussion you found! Now, the challenge is even harder. The existing behavior is probably acceptable in ffmpeg because when **lame** crashes due to garbage input, this is only the current ffmpeg session. With **mobile-ffmpeg**, the whole mobile app crashes, which is a whole different user experience.
> 

Ideally, if we could run tests, it would help us to see the impact of settings this flag. Without tests, I can not foresee whether setting this flag will introduce new crashes or not. Honestly, I'm not worried about generating garbage output. Stability and memory usage are my main concerns here.

> As Alexander Leidinger [recently explained](https://sourceforge.net/p/lame/bugs/491/#82db/a8be/eebf), one cannot simply skip the `asserts` in **lame**. What can probably be done, is redefine their behavior and cancel current execution instead of crashing the app. But if this is not reported too often (whatever your definition of _too often_), this fix can wait.
> 
> You still may gain from enabling `NDEBUG` for other libraries. Note that this is a standard behavior for all build systems, including NDK and XCode.

I think `NDEBUG` is already appended by other libraries that support it. `./configure` is responsible of of setting those flags and if we're not creating a debug build, then it should have been set. This case about `lame` happens because it was explicitly commented out.

**Comment by alexcohn** [comment]
> `./configure` is responsible of setting those flags

This makes sense. I don't really see it in many libraries, but in **snappy** for example, it is set. But that's right, when there is official `.configure`, this should be their responsibility.

> Ideally, if we could run tests, it would help us to see the impact of setting this flag.

If **lame** project cannot have this tested and ensured, it is certainly out of scope of **mobile-ffmpeg**. On the other hand, it may be possible to change the behavior of `assert()` to protect the whole app from crashing.

**Comment by tanersener** [comment]
Well, `lame` does have tests. `Makefile` has a `test` target. But we can't run those tests on iOS or Android. We can run them on build host only, which won't help us much. It is a limitation of cross-compilation tools.

And, overriding or changing the behaviour of `assert()` won't be easy. `assert()` is called inside `lame` and we need to notify `mobile-ffmpeg` when `assert()` is called/triggered. Too complex to solve.

Setting `NDEBUG` is still the easiest option. Setting the flag for one of the `Main` releases and monitoring the issues might be a more realistic option.

**Comment by alexcohn** [comment]
> Well, `lame` does have tests.

but even with these tests, they rely on `assert()` to ensure security (at the price of stability).

Actually, it should be entirely possible to change the behavior of `assert()` in `lame` so that it interrupts the current `Config.ffmpegExecute()` only. I can provide a POC next week. Unlike the potential dangers of buffer overflow when the `assert()` is ignored, this can also be fully tested.

**Comment by tanersener** [comment]
Are you sure it is worth working on? This issue is raised only once in two years.

**Comment by alexcohn** [comment]
No, I am not sure it's worth it. But hiding the failure by disabling asserts is not the way to deal with it.
